### DISCUSSION TITLE
Overriding invariance of variadic generics by subclassing type alias

### DISCUSSION THREAD

**Discussion by sylee957** [microsoft/pyright#4841] [Q&A] [answered]
I have an example where invariance of variadic generics can be overridden

There are 

- A variadic generic `Invariant`
- A simple subtyping relationship between `Animal` and `Dog` which is unrelated to the `Invariant`

used in the example

```python3
from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, TypeAlias, TypeVar, TypeVarTuple, reveal_type

T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)
Ts = TypeVarTuple("Ts")


@dataclass(init=False)
class Invariant(Generic[*Ts]):
    args: tuple[*Ts]

    def __init__(self, *args: *Ts):
        self.args = args


class Animal:
    pass


class Dog(Animal):
    pass
```

Obviously, this doesn't passes the type checks because the `TypeVarTuple` is assumed invariant 
```python3
a: Invariant[Animal] = Invariant(Animal())
b: Invariant[Dog] = Invariant(Dog())
a = b
```

And this doesn't passes the type check because of the error `Variance of type argument is incompatible`

```python
class Covariant(Invariant[T_co]):
    pass
```

However, when I start subclassing the type alias, it works.

```python
_Covariant: TypeAlias = Invariant[T]


class Covariant(_Covariant[T_co]):
    pass


a: Covariant[Animal] = Covariant(Animal())
b: Covariant[Dog] = Covariant(Dog())
a = b
```

Using `_Covariant: TypeAlias = Invariant[T_co]` also works.
I'm not sure if this is inconsistent behavior, or can be a trick to deal with the unsupported variance of `TypeVarTuple` for some cases.

**Comment by erictraut** [comment, marked as answer]
Thanks for the bug report. This will be addressed in the next release.
